### DISCUSSION TITLE
Support for Pre-order items

### DISCUSSION THREAD

**Discussion by herobaby71** [saleor/saleor#12070] [General] [answered]
Hi guys,

**Pre-order flow in our previous store:**
We have pre-order products (up to 12 months lag time) on our webstore and need to charge 10% upfront. To do this, we plan to add a category 'Pre-Order' to the Products, and apply a 90% discount voucher for these 'Pre-Order' items during checkout. After the items arrive, we can go to Saleor dashboard and remove the voucher to display the remaining amount in the storefront, which the user can then login and pay.

I have a few questions:

First, do you guys know if Saleor has a feature available for Pre-Order since I could not find the solution anywhere on the documentations and discussions?

Second, we cannot find a way to remove or edit the order line in the dashboard so we cannot perform the described flow above. Do you guys know if there is a way to enable such feature?

Third, do you know any potential problems with such pre-order flow because we are currently at the early stage of development?

Cheers,

**Comment by korycins** [comment, marked as answer]
Hey @herobaby71 
I would suggest trying the orderCreateFromCheckout mutation, and the transactionItem(s). Transaction flow is our new way of handling payment by Saleor (including split payments) (introduced in 3.4).
You could have your checkout app, that communicates with Saleor, and create an order that is not fully paid/not paid (by using protected orderCreateFromCheckout mutation), then you can attach multiple transactionItems. In your case you could probably have:
- 1st transaction, 10% of the total amount
- 2nd transaction, with 90% attached when products are ready to send and the user covered the rest of the order total.

The transaction flow is in feature preview mode, and you need to take into account that something can be changed there across future versions but it sounds like something perfect for you: the possibility to create an order that is not fully paid, attaching multiple payments (1st at the beginning, second after x days).
The docs that describe the flow: https://docs.saleor.io/docs/3.x/developer/checkout#finalizing-checkout-with-saleor-app
The changes that we are going to introduce soon: https://github.com/saleor/saleor-docs/pull/659 ([link to preview docs](https://saleor-docs-git-update-payment-flow-with-f89ba6-saleorcommerce.vercel.app/docs/3.x/developer/payments)) These changes will also introduce a dashboard support for transactions
The changes that we are implementing right now: https://github.com/saleor/saleor/issues/11258

**Reply by herobaby71** [reply]
Thank you @korycins. I have some questions and want to make sure I understand the flow here.

Assuming the customer has a Checkout with a CheckoutLine with a pre-order metadata flag, the client will first Invoke the Saleor checkout app to create an Order from the Checkout via the orderCreateFromCheckout [1]. 

After converting the user Checkout to Order, the client then invokes the Saleor checkout app again to create a payment [2]. Once the Saleor checkout app received the payment, the app then creates two transactions attached to Order with different amountAuthorized and amountCharged [3].

My questions are:
1/ Since the customer is being charged at the create a payment stage [2], does this mean I have to customise the saleor checkout app to check the Order metadata for pre-order status and change the Payment amount appropriately?

2/ I have been assuming that the created Order from step [1] has the pre-order metadata. However, do you know if the orderCreateFromCheckout mutation creates an Order with the Orderline with similar metadata to that of the CheckoutLine?

3/ Are there any rules or connections between the Payment's amount and the Transaction `amountAuthorized` and `amountCharged` in the Transactions, or can I change this freely?
Cheers

**Reply by mociepka** [reply]
Hi, @herobaby71 you can find it at the API docs:

https://docs.saleor.io/docs/3.x/api-reference/mutations/transaction-create.

**Reply by herobaby71** [reply]
Hi @mociepka , I edited my response above after reading the documentations. Can you help me check whether my understanding is correct? Cheers

**Reply by korycins** [reply]
Hey @herobaby71 

> Assuming the customer has a Checkout with a CheckoutLine with a pre-order metadata flag, the client will first Invoke the Saleor checkout app to create an Order from the Checkout via the orderCreateFromCheckout [1].

You can perform the following sequence of actions: first, place the order; then, make the initial payment; and finally, make the second payment.

Alternatively, you can use the transaction API and orderCreateFromCheckout to follow this flow: make the initial payment, then place the order, and finally make the second payment.


> After converting the user Checkout to Order, the client then invokes the Saleor checkout app again to create a payment [2]. Once the Saleor checkout app received the payment, the app then creates two transactions attached to Order with different amountAuthorized and amountCharged [3].

I recommend creating a single transaction with 10% of the total order amount as the amountCharged, as this is the amount captured from the customer. After X days, when the pre-order is ready to be sent, the customer can pay for the remaining amount of the order. This will create a second transaction with 90% of the order total as the amountCharged.

>  Since the customer is being charged at the create a payment stage [2], does this mean I have to customise the saleor checkout app to check the Order metadata for pre-order status and change the Payment amount appropriately?

Correct. The Checkout App supports a basic flow to demonstrate how the new interface works. However, if you have custom logic that needs to be executed before creating an order, you will need to customize the Checkout App or build a new one, using our Checkout App as an example.

>  I have been assuming that the created Order from step [1] has the pre-order metadata. However, do you know if the orderCreateFromCheckout mutation creates an Order with the Orderline with similar metadata to that of the CheckoutLine?

By default, all metadata fields from the checkout are migrated to the order.

> Are there any rules or connections between the Payment's amount and the Transaction amountAuthorized and amountCharged in the Transactions, or can I change this freely?

Do you mean the `Payment` object that is created for the old payment flow when you write "Payment"?

Please note that we do not allow mixing the old payment flow with the new transaction flow within a single checkout/order.

The transaction amounts are used to calculate the current payment status of the order. You can modify the amounts by calling the `transactionUpdate` mutation or the `transactionEventReport` mutation (3.12, the recommended way of updating the transaction). Starting from version 3.12, each call to `transactionUpdate` or `transactionEventReport` will create a new `transactionEvent` object with the difference between the amount you provided and the amount stored in the transaction. This allows you to track any changes that have happened on the given transaction.

For example, let's say you have an order for 100 USD, but you don't have auto-capture enabled. When the customer pays for the order, only 100 USD is authorized (as you have disabled auto-capture on the payment provider side), and the transaction related to this action should have `authorizedAmount` set to 100 USD. When a staff user decides to call capture, the `chargedAmount` should be updated for the given transaction, either by using `transactionEventReport` (recommended) or by `transactionUpdate`.

**Reply by herobaby71** [reply]
Thank you for your detailed response @korycins, I will take a deeper look into the APIs in the upcoming weeks and will get back if something is unclear. Cheers!
